alphagov/paas-cf
# [#120846679] Feature flag destroy pipelines off in non-dev envs

## Notice

Before merging we need to capture logs from the `destroy-cloudfoundry` pipeline in prod. I'm not able to do this at the moment, so if somebody else could..

## What

Add a new `ENABLE_DESTROY` feature flag to control whether the
`destroy-cloudfoundry` and `destroy-microbosh` pipelines are setup on the
Deployer Concourse. This is enabled on development environments and disabled
by default on all other environments such as CI, staging, and production.
Instructions for enabling the destroy pipelines in these environments, if
necessary, is added to the non-dev deployments documentation.

This will prevent the destroy pipelines from being run by accident in
persistent environments. We've seen this happen once and it's easy to
do when you have multiple Concourse tabs open to diagnose a pipeline failure
in another environment.

I've not included a warning like `ENABLE_AUTODELETE` does because it will be
quite noisy and the effect isn't as surprising as autodelete; which contains
a pipeline that runs automatically. I've used a glob to capture both destroy
pipelines in the case statement so that we don't have to match their exact
names. These pipelines aren't reinstated by `self-update-pipeline` because
that passing in a single pipeline name.

We have talked about adding a Concourse feature to make the background a
different colour for different environments or adding some text to the
pipeline screen. Those things might still be desirable, but we should leave
this in place for extra safety.

## How to review

1. Checkout the branch: `git fetch && git checkout 120846679-feature_flag_destroy_pipelines`
1. Update the pipelines on your dev environment: `make dev pipelines`
1. Confirm that the destroy pipelines are still on the Deployer Concourse.
1. Temporarily delete the `ENABLE_DESTROY=true` line from the `Makefile` to simulate a non-dev environment.
1. Update the pipelines again: `make dev pipelines`
1. Confirm that the destroy pipelines are *not* on the Deployer Concourse.
1. Run the `create-bosh-cloudfoundry` and confirm that the `init` job doesn't reinstate them.

## Who can review

Not @dcarley
